src/app/modules/myinfo/myinfo.util.ts

```diff
@@ -0,0 +1,315 @@
+        errorMessage:
+          'Could not find the form requested. Please refresh and try again.',
+      }
+    case MyInfoAuthTypeError:
+      return {
+        statusCode: StatusCodes.BAD_REQUEST,
+        errorMessage:
+          'This form does not have MyInfo enabled. Please refresh and try again.',
+      }
+    case MyInfoNoESrvcIdError:
+      return {
+        statusCode: StatusCodes.FORBIDDEN,
+        errorMessage: 'This form does not have a valid e-service ID.',
+      }
+    case FetchLoginPageError:
+    case LoginPageValidationError:
+      return {
+        statusCode: StatusCodes.SERVICE_UNAVAILABLE,
+        errorMessage: 'Failed to contact SingPass. Please try again.',
+      }
+    case DatabaseError:
+    case MissingFeatureError:
+    case CreateRedirectUrlError:
+      return {
+        statusCode: StatusCodes.INTERNAL_SERVER_ERROR,
+        errorMessage: coreErrorMessage,
+      }
+    default:
+      logger.error({
+        message: 'Unknown route error observed',
+        meta: {
+          action: 'mapRedirectURLError',
+        },
+        error,
+      })
+      return {
+        statusCode: StatusCodes.INTERNAL_SERVER_ERROR,
+        errorMessage: coreErrorMessage,
+      }
+  }
+}
+
+export const getMyInfoFieldOptions = (
+  myInfoAttr: IMyInfo['attr'],
+): string[] => {
+  const [myInfoField] = myInfoTypes.filter((type) => type.name === myInfoAttr)
+  return myInfoField?.fieldOptions || []
+}
+
+export const createConsentPagePurpose = (formTitle: string): string =>
+  `The form "${formTitle}" is requesting to pre-fill your MyInfo data.`
+
+export const createRelayState = (formId: string, rememberMe: boolean): string =>
+  `${uuidv4()},${formId},${rememberMe}`
+
+export const validateMyInfoForm = (
+  form: IFormSchema | IPopulatedForm,
+): Result<IMyInfoForm, MyInfoNoESrvcIdError | MyInfoAuthTypeError> => {
+  if (!form.esrvcId) {
+    return err(new MyInfoNoESrvcIdError())
+  }
+  if (form.authType !== AuthType.MyInfo) {
+    return err(new MyInfoAuthTypeError())
+  }
+  return ok(form as IMyInfoForm)
+}
+
+const hasProp = <K extends string>(
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  obj: object | Record<string, unknown>,
+  prop: K,
+): obj is Record<K, unknown> => {
+  return prop in obj
+}
+
+export const extractMyInfoCookie = (
+  cookies: Record<string, unknown>,
+): Result<MyInfoCookiePayload, MyInfoMissingAccessTokenError> => {
+  const cookie = cookies[MYINFO_COOKIE_NAME]
+  if (cookie && typeof cookie === 'object' && hasProp(cookie, 'state')) {
+    if (
+      cookie.state === MyInfoCookieState.AccessTokenRetrieved &&
+      hasProp(cookie, 'accessToken') &&
+      typeof cookie.accessToken === 'string' &&
+      hasProp(cookie, 'usedCount') &&
+      typeof cookie.usedCount === 'number'
+    ) {
+      return ok(cookie as MyInfoCookiePayload)
+    } else if (cookie.state === MyInfoCookieState.RetrieveAccessTokenError) {
+      return ok(cookie as MyInfoCookiePayload)
+    }
```

Comment:
recommend extracting the logic into a type guard, instead of explicit typecasting to `MyInfoCookiePayload`. would it be beneficial to explicitly distinguish between success and failure types?

Reply:
re: distinguishing between success and failure types, I simplified the states to just success and error because it doesn't matter what the error state is. so I don't think there's a need to distinguish between the types - you just need to check that the state is `Success`.